File: 0256-paint-house/0256-paint-house.cpp
class Solution {
    
    int minCost(vector<vector<int>>& costs, int index, int prev, vector<vector<int>>& dp)
    {
        if(index == costs.size())
            return 0;
        
        if(dp[index][prev + 1] != -1)
            return dp[index][prev + 1];
        
        int cost = INT_MAX;
        for(int i=0; i<3; i++)
        {
            if(prev == -1 or i != prev)                
                cost = min(cost, costs[index][i] + minCost(costs, index + 1, i, dp)); 
        }
    return dp[index][prev + 1]= cost;
    }
    
public:
    int minCost(vector<vector<int>>& costs) {
        vector<vector<int>>dp(costs.size(), vector<int>(4, -1));
        return minCost(costs, 0, -1, dp);
    }
};

File: 0256-paint-house/0256-paint-house.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        
        dp = {} 
        def cost(i, prev):
            
            if i == len(costs):
                return 0
            if (i, prev) in dp:
                return dp[i, prev]
            
            res = float('inf')
            for color in range(0, 3):
                
                if prev == -1 or color != prev:
                    res = min(res, costs[i][color] + cost(i + 1, color))
            dp[i, prev] = res
            return res
                    
        return cost(0, -1)
                